IQSS/dataverse
# Externally controlled vocabulary support

Contributors:

- @4tikhonov 
- @ekoi 
- @PaulBoon 

**What this PR does / why we need it**: adds support for externally controlled vocabularies to compound-fields with the following four subfields: Vocabulary, Vocabulary URL, Term, Term URL. The vocabulary server must provide a SKOSMOS
interface.

**Which issue(s) this PR closes**:

Closes #7711 

**Special notes for your reviewer**: 

Known problems/open issue:

* JSON configuration lay-out should be cleaned up, bundling related settings into objects.
* Add ability to configure a vocabulary URL per vocabulary.
* If a vocabulary is pre-filled, then it is not possible to leave the entire compound field empty, because the vocabulary URL is always filled in.
* If you select a term from the list, the Term URL is automatically filled in. After that you can change the term, resulting in a term that does not correspond with the term URL.

**Suggestions on how to test this**:

How to test this

1. Clone, built and deploy this war file
2. Apply sql statements to update Dataverse database: 
`psql dvndb -f cv-update.sql`
Warning: dvndb is postgresql database, it can be different in your installation.
[cv-update.sql.zip](https://github.com/DANS-KNAW/dataverse/files/6188821/cv-update.sql.zip)
3. Add this custom metadata block (cvocdemo.tsv) to Dataverse
See: http://guides.dataverse.org/en/latest/admin/metadatacustomization.html for an explanation of the steps to perform. 
[cvocdemo.tsv.zip](https://github.com/DANS-KNAW/dataverse/files/6188466/cvocdemo.tsv.zip)
4. Enable the 'cvocDemo Metadata' Metadata Block for the dataverse where you are creating the dataset to test with
5. Configure Dataverse to use the external-cvoc feature on the metdata block. The configuration is in a json file (cvoc-conf.json). Run the following command to load the config into Dataverse

    ```bash 
    curl -H "Content-Type: application/json" -X PUT \ 
              -d @cvoc-conf.json http://localhost:8080/api/admin/settings/:CVocConf
    ```
    [cvoc-conf.json.zip](https://github.com/DANS-KNAW/dataverse/files/6188527/cvoc-conf.json.zip) You can change the config and reload it to try different settings. For instance using only the unesco vocabulary with `"vocabs":["unesco"],`    instead of `"vocabs":["unesco","stw","agrovoc"],`.

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:

The Dataset creation form will now contain the following extra block. Without the configuration to connect it to the controlled vocabulary service it looks like this:
![Screenshot cvocDome-empty](https://user-images.githubusercontent.com/2151103/112136800-bff61a00-8bcf-11eb-9656-5df48e09c808.png)

With one vocabulary (unesco), it looks like this:
![Screenshot cvocDemo-one-vocab](https://user-images.githubusercontent.com/2151103/112136911-e74ce700-8bcf-11eb-8258-0e8dcd202b58.png)

With several vocabularies:
![Screenshot cvocDemo-multi-vocabs](https://user-images.githubusercontent.com/2151103/112137009-05b2e280-8bd0-11eb-9770-227d278dbfc0.png)

**Is there a release notes update needed for this change?**:  ?

**Additional documentation**:

Design

![image](https://user-images.githubusercontent.com/3862694/112156525-9ac0d600-8be6-11eb-8d49-94a75dff4968.png)

The cvoc-enabled compound field communicates via the SKOSMOS protocol with a vocabulary server to retrieve matching vocabularies, terms and their correspoding URIs.

Example config

```json
[
   {   
     "vocab-name":"cvocDemo",
     "minChars": 1,
     "cvoc-url":"https://skosmos.dev.finto.fi/",
     "language":"en",
     "js-url":"/resources/js/cvoc-interface.js",
     "protocol": "skosmos",
     "vocab-uri" : "http://skos.um.es/unescothes/CS000", 
     "term-parent-uri": "",
     "vocabs":["unesco","stw","agrovoc"],
     "vocab-codes": ["cvocDemoVocabulary", "cvocDemoTerm", "cvocDemoTermURI", "cvocDemoVocabularyURI"]
   }
]
```

Description of the configuration settings

* vocab-name
  Name of the compound metadata field (Need to check usage in the code)
  
* cvoc-url
URL of the external vocabulary service that will provide the terms and URI's in the autocomplete
Must now be a SKOSMOS service, because that API is the only one supported by the cvoc-interface.js code; see "js-url" property. 
   
* language
Specifies the language for the terms (prefLabel in SKOS) to be retrieved. 
By default it's using the language of the server (calling BundleUtil.getDefaultLocale().getLanguage())

* js-url
Optional URL of the JavaScript file that does the Ajax request. A default JavaScript is included in Dataverse.
  
* protocol
Note that only "skosmos" is supported in the cvoc-interface.js. The default is also "skosmos".
  
* vocab-uri
The URI of the vocab, this is now only a single value but should be an array corresponding with the "vocabs" array.  (Note that it is an URL we want to put in that input field)
  
* term-parent-uri
The URI of the 'parent' concept acting as a filter for the terms to be retrieved allowing only narrower terms. This enables us to use a 'branch' of a large vocabulary without constructing separate vocabularies for subsets.    
  
* vocabs
The list of vocabularies. This should be the 'identifying' names in SKOSMOS. 

* vocab-codes
The input field names as specified in the metadata block. The order of the fields must be; vocabulary, term, termURI, VocabularyURI. 

* minChars
  Number of characters that must be typed in the 'term' field for the AJAX 'search' request to be sent. Use this with large vocabularies and set it to more characters (1 is usually enough) if the server does not respond fast enough. When it is 0 (default), the complete list will dropdown when the up or down arrow keys are typed. 
  
* readonly
Specify if the URL fields should be readonly. Also the vocabulary field is readonly when there is only one vocab configured. The default is false, allowing to type anything into the fields.   
  
* hideReadonlyUrls
When in 'readonly' the URL fields are now also hidden on the form. This makes the form more compact and simpler. The default is false, showing all the fields.
